devito/petsc/iet/passes.py

```diff
@@ -0,0 +1,402 @@
+from collections import OrderedDict
+import cgen as c
+
+from devito.passes.iet.engine import iet_pass
+from devito.ir.iet import (FindNodes, Call,
+                           Transformer, FindSymbols,
+                           MapNodes, Iteration, Callable, Callback, List, Uxreplace,
+                           Definition, BlankLine, PointerCast, filter_iterations,
+                           retrieve_iteration_tree)
+from devito.symbolics import Byref, Macro, FieldFromPointer
+from devito.petsc.types import (PetscMPIInt, PETScStruct, DM, Mat,
+                                Vec, KSP, PC, SNES, PetscErrorCode, PETScArray)
+from devito.petsc.iet.nodes import MatVecAction, LinearSolverExpression
+from devito.petsc.utils import (solver_mapper, petsc_call, petsc_call_mpi,
+                                core_metadata)
+
+
+@iet_pass
+def lower_petsc(iet, **kwargs):
+
+    # Check if PETScSolve was used
+    petsc_nodes = FindNodes(MatVecAction).visit(iet)
+
+    if not petsc_nodes:
+        return iet, {}
+
+    # Collect all petsc solution fields
+    unique_targets = list({i.expr.rhs.target for i in petsc_nodes})
+    init = init_petsc(**kwargs)
+
+    # Assumption is that all targets have the same grid so can use any target here
+    objs = build_core_objects(unique_targets[-1], **kwargs)
+    objs['struct'] = build_petsc_struct(iet)
+
+    # Create core PETSc calls (not specific to each PETScSolve)
+    core = make_core_petsc_calls(objs, **kwargs)
+
+    # Create matvec mapper from the spatial iteration loops (exclude time loop if present)
+    spatial_body = []
+    for tree in retrieve_iteration_tree(iet):
+        root = filter_iterations(tree, key=lambda i: i.dim.is_Space)[0]
+        spatial_body.append(root)
+    matvec_mapper = MapNodes(Iteration, MatVecAction,
+                             'groupby').visit(List(body=spatial_body))
+
+    setup = []
+
+    # Create a different DMDA for each target with a unique space order
+    unique_dmdas = {}
+    for target in unique_targets:
+        name = 'da_so_%s' % target.space_order
+        unique_dmdas[name] = DM(name=name, liveness='eager',
+                                stencil_width=target.space_order)
+    objs.update(unique_dmdas)
+    for dmda in unique_dmdas.values():
+        setup.extend(create_dmda_calls(dmda, objs))
+
+    subs = {}
+    efuncs = OrderedDict()
+
+    # Create the PETSc calls which are specific to each target
+    for target in unique_targets:
+        solver_objs = build_solver_objs(target)
+        matvec_body_list = List()
```

Comment:
instead of creating a List and rebuilding it each time you have to add something, why don't you just append to a python list, and create the devito List at the very end?